### PR TITLE
snapstate: allow `snap refresh` to refresh even if the snap is already installed

### PR DESCRIPTION
**Pull request by mvo5**
This allows to do something like:
```
$ sudo snap install --channel=stable foo
Name Rev
foo  1
$ sudo snap refresh --channel=beta foo
Name Rev
foo  2
$ sudo snap refresh --channel=stable foo
Name Rev
foo  1
```
I.e. to refresh to a revision that is already installed locally.

This also closes: LP:#1594330 which is a different incarnation of the same underlying issue.

### CODE REVIEW

File: overlord/snapstate/snapmgr.go [hunk 1]
```diff
@@ -325,13 +325,19 @@ func (m *SnapManager) doDownloadSnap(t *state.Task, _ *tomb.Tomb) error {
...
+	if snapst.Current == storeInfo.Revision {
+		return fmt.Errorf("revision %s of snap %q already current", storeInfo.Revision, ss.Name)
 	}
 
-	downloadedSnapFile, err := m.store.Download(storeInfo, meter, auther)
-	if err != nil {
-		return err
+	var downloadedSnapFile string
+	if revisionIsLocal(ss.Name, snapst, storeInfo.Revision) {
+		downloadedSnapFile = storeInfo.MountFile()
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go]: this feels quite like a hack tbh

**Comment by pedronis** [comment]
also I'm not sure this does the right thing vs copying data, because it will copy it over the old data from the revision which might or might not be the right thing

**Comment by pedronis** [comment]
we also have the pending issue of making refresh actually different from install in terms of using the right store apis, not sure what should be done first...

**Comment by mvo5** [comment]
Closing this too as it needs some serious rework (i.e. we need to talk to the store earlier).

**Pull request closed without merging**